sdk/Lusid.Sdk.Tests/Utilities/PollyApiRetryHandlerTests.cs

```diff
@@ -0,0 +1,409 @@
+        {
+            const int returnedStatusCode = 502;
+            const int expectedNumberOfRetries = 2;
+            for (var i = 0; i < expectedNumberOfRetries + 1; i++)
+                // Every response fails
+                AddMockHttpResponseToQueue(_httpListener, statusCode: returnedStatusCode,
+                    responseContent: "Error that was thrown");
+            var retryCount = 0;
+
+            RetryConfiguration.RetryPolicy =
+                Policy
+                    .HandleResult<IRestResponse>(response => response.StatusCode == (HttpStatusCode) returnedStatusCode)
+                    .Retry(expectedNumberOfRetries, onRetry: (result, i) => retryCount++);
+
+            // Calling GetPortfolio or any other API triggers the flow that triggers polly
+            var response = _apiFactory.Api<IPortfoliosApi>().GetPortfolio("any", "any");
+
+            Assert.That(retryCount, Is.EqualTo(expectedNumberOfRetries));
+            // Todo: In the future 0 error codes with throw an error after retries exceeded
+            Assert.That(response, Is.Null);
+        }
+
+        [Test]
+        public void CallGetPortfoliosApi_PollyRetryConditionIsSatisfied_RetriesUntilSuccess_DoesNotThrow()
+        {
+            const int returnedStatusCode = 499; // Or any custom defined code
+            const int expectedNumberOfRetries = 2;
+            // First Response is a failing code
+            AddMockHttpResponseToQueue(_httpListener, statusCode: returnedStatusCode, responseContent: "");
+            // Second response is the first retry. It fails again
+            AddMockHttpResponseToQueue(_httpListener, statusCode: returnedStatusCode, responseContent: "");
+            // Third response is the second retry. Returns 200 before retries are exceeded, and does not throw
+            AddMockHttpResponseToQueue(_httpListener, statusCode: 200, responseContent: "");
+            var retryCount = 0;
+            RetryConfiguration.RetryPolicy = Policy.Wrap(
+                PollyApiRetryHandler.DefaultFallbackPolicy,
+                Policy
+                    .HandleResult<IRestResponse>(response => response.StatusCode == (HttpStatusCode) returnedStatusCode)
+                    .Retry(expectedNumberOfRetries, onRetry: (result, i) => retryCount++));
+
+            // Calling GetPortfolio or any other API triggers the flow that triggers polly
+            var sdkResponse = _apiFactory.Api<IPortfoliosApi>().GetPortfolio("any", "any");
+
+            Assert.That(retryCount, Is.EqualTo(expectedNumberOfRetries));
+            // Todo: In the future 0 error codes with throw an error after retries exceeded
+            Assert.That(sdkResponse, Is.Null);
+        }
+
+        // Example of how an exponential backoff can be used with Polly
+        [Test]
+        public void CallGetPortfoliosApi_WhenApiResponseStatusCodeSatisfiesRetryCriteria_PollyRetryWithBackoffIsTriggered()
```

Comment:
Is this more of a tutorial than a test?

Reply:
Yes, since this is a client facing code and the proper way of implementing polly is tricky due to OpenAPI code, I think it's good to have these examples